### ISSUE TITLE
Different debug and pre hooks firing sequence/order for find() and aggregate()

### ISSUE THREAD

**Issue by C-Duv** [Automattic/mongoose#7606]
**Do you want to request a *feature* or report a *bug*?**

Report a possible bug 

**What is the current behavior?**

When using `debug` and `pre` hooks I get different firing sequence if query is a `find()` or an `aggregate()`.

In both case (either `find()` or `aggregate()`), I get the following order:
1. pre-hook
2. debug
3. post-hook

But the properties (eg. `options`) I add the query object in `pre`-hook are not available to `debug` when query is an `aggregate()` (but are with `find()`).
So it looks like, for `aggregate()`, `debug` gets fired in parallel of `pre`.

```Js
mongoose.set('debug', (collectionName, method, query, queryOptions) => {
    console.log(`[${new Date().toISOString()}] debug-mongoose: query (${method}):`, query);
    console.log(`[${new Date().toISOString()}] debug-mongoose: queryOptions:`, queryOptions);

    if (
        queryOptions.hasOwnProperty('profiler')
        && queryOptions.profiler.hasOwnProperty('startTime')
    ) {
        console.log(`[${new Date().toISOString()}] debug-mongoose: query started at ${queryOptions.profiler.startTime.toISOString()}`);
    }
    if (
        queryOptions.hasOwnProperty('profiler')
        && queryOptions.profiler.hasOwnProperty('endTime')
    ) {
        console.log(`[${new Date().toISOString()}] debug-mongoose: query ended at ${queryOptions.profiler.endTime.toISOString()}`);
    }
});

const preHook = function() {
    this.options.profiler = {
        'startTime': new Date(),
    };
    console.log(`[${new Date().toISOString()}] pre-hook: query started at ${this.options.profiler.startTime.toISOString()}`);
    console.log(`[${new Date().toISOString()}] pre-hook: this.options:`, this.options);
};
const postHook = function(result) {
    this.options.profiler.endTime = new Date();
    console.log(`[${new Date().toISOString()}] post-hook: query ended at ${this.options.profiler.endTime.toISOString()}`);
    console.log(`[${new Date().toISOString()}] post-hook: this.options:`, this.options);
}
```

Find outputs (where `debug-mongoose: queryOptions` line shows `profiler` query option set by `pre` is available in `debug` hook):
```
[2019-03-14T10:27:55.411Z] pre-hook: query started at 2019-03-14T10:27:55.411Z
[2019-03-14T10:27:55.411Z] pre-hook: this.options: { profiler: { startTime: 2019-03-14T10:27:55.411Z } }
[2019-03-14T10:27:55.414Z] debug-mongoose: query (find): { foo: 42 }
[2019-03-14T10:27:55.415Z] debug-mongoose: queryOptions: { profiler: { startTime: 2019-03-14T10:27:55.411Z },
  projection: {} }
[2019-03-14T10:27:55.415Z] debug-mongoose: query started at 2019-03-14T10:27:55.411Z
[2019-03-14T10:27:55.424Z] post-hook: query ended at 2019-03-14T10:27:55.424Z
[2019-03-14T10:27:55.424Z] post-hook: this.options: { profiler:
   { startTime: 2019-03-14T10:27:55.411Z,
     endTime: 2019-03-14T10:27:55.424Z } }
```

Aggregate outputs (where `debug-mongoose: queryOptions` line shows `profiler` query option set by `pre` is not available in `debug` hook):
```
[2019-03-14T10:27:34.742Z] pre-hook: query started at 2019-03-14T10:27:34.742Z
[2019-03-14T10:27:34.742Z] pre-hook: this.options: { profiler: { startTime: 2019-03-14T10:27:34.742Z } }
[2019-03-14T10:27:34.747Z] debug-mongoose: query (aggregate): [ { '$project': { a: 1, b: 1 } }, { '$limit': 1 } ]
[2019-03-14T10:27:34.748Z] debug-mongoose: queryOptions: {}
[2019-03-14T10:27:34.763Z] post-hook: query ended at 2019-03-14T10:27:34.763Z
[2019-03-14T10:27:34.763Z] post-hook: this.options: { profiler:
   { startTime: 2019-03-14T10:27:34.742Z,
     endTime: 2019-03-14T10:27:34.763Z } }
```

Here is a full gist to reproduce:
https://gist.github.com/C-Duv/9d547f951f0ee94e12b8c36d6bd4bf24

**What is the expected behavior?**

Consistent behavior between `find()` and `aggregate()`.

**Please mention your node.js, mongoose and MongoDB version.**

* NodeJS: v10.15.3
* Mongoose: v5.4.17
* MongoDB: v4.0.3

**Comment by vkarpov15** [comment]
Thanks for pointing out this issue, turns out you couldn't set options in pre('aggregate') at all. Fix will be in 5.4.22 :+1:

**Comment by C-Duv** [comment]
Thanks: I can confirm it now works :+1:

```Shell
cd $(mktemp -d)
git clone https://gist.github.com/9d547f951f0ee94e12b8c36d6bd4bf24.git .
npm install
npm view mongoose version
# 5.4.22
node mongoose-debug_pre_post_hooks_order-reproduction-code.js
# [2019-04-08T09:51:39.019Z] pre-hook: query started at 2019-04-08T09:51:39.019Z
# [2019-04-08T09:51:39.021Z] pre-hook: this.options: { profiler: { startTime: 2019-04-08T09:51:39.019Z } }
# [2019-04-08T09:51:39.028Z] pre-hook: query started at 2019-04-08T09:51:39.028Z
# [2019-04-08T09:51:39.028Z] pre-hook: this.options: { profiler: { startTime: 2019-04-08T09:51:39.028Z } }
# [2019-04-08T09:51:39.029Z] debug-mongoose: query (aggregate): [ { '$project': { a: 1, b: 1 } }, { '$limit': 1 } ]
# [2019-04-08T09:51:39.030Z] debug-mongoose: queryOptions: { profiler: { startTime: 2019-04-08T09:51:39.019Z } }
# [2019-04-08T09:51:39.030Z] debug-mongoose: query started at 2019-04-08T09:51:39.019Z
# [2019-04-08T09:51:39.035Z] debug-mongoose: query (find): { foo: 42 }
# [2019-04-08T09:51:39.039Z] debug-mongoose: queryOptions: { profiler: { startTime: 2019-04-08T09:51:39.028Z },
#   projection: {} }
# [2019-04-08T09:51:39.040Z] debug-mongoose: query started at 2019-04-08T09:51:39.028Z
# [2019-04-08T09:51:39.048Z] post-hook: query ended at 2019-04-08T09:51:39.048Z
# [2019-04-08T09:51:39.049Z] post-hook: this.options: { profiler:
#    { startTime: 2019-04-08T09:51:39.019Z,
#      endTime: 2019-04-08T09:51:39.048Z } }
# [2019-04-08T09:51:39.054Z] post-hook: query ended at 2019-04-08T09:51:39.054Z
# [2019-04-08T09:51:39.054Z] post-hook: this.options: { profiler:
#    { startTime: 2019-04-08T09:51:39.028Z,
#      endTime: 2019-04-08T09:51:39.054Z } }
^C
rm -Rf $OLDPWD
```
